### PR TITLE
Add Volumetric Cloud Compatibility with MSAA [Recreated]

### PR DESCRIPTION
**Pull request by johnpars**
### Purpose of this PR

Fix case: https://fogbugz.unity3d.com/f/cases/1327436/

Volumetric Clouds will fail in the case of MSAA color target. This is due to the in-place blending operation that occurs on the upscaling + combine pass (which can't be done on an MSAA target). This PR will move that combine to a subsequent pass after the upscaling, and instead do the blending via the hardware on the MSAA target. 

Volumetric clouds now work in both scenarios:

No MSAA:
image

MSAA:
image

---
### Testing status
- I locally tested switching the number of MSAA samples on different graphics APIs. 
- I also ensured that the hardware blending operation produces the exact same result as when it is done in the CS pass. 
- [Yamato](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Ffix-1327436)

---
### Comments to reviewers
Note: This PR is a recreation of #4345, which got mangled due do a deletion of hd/bugfix. The branch was also pretty far behind master (and volumetric clouds received various updates in the meantime), so I did a fresh branch off from master and cherry picked the work from the original branch.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs [hunk 1]
```diff
@@ -43,6 +46,8 @@ void InitializeVolumetricClouds()
             m_UpscaleAndCombineCloudsKernelColorCopy = volumetricCloudsCS.FindKernel("UpscaleAndCombineClouds_ColorCopy");
             m_UpscaleAndCombineCloudsKernelColorRW = volumetricCloudsCS.FindKernel("UpscaleAndCombineClouds_ColorRW");
 
+            m_CloudCombinePass = CoreUtils.CreateEngineMaterial(defaultResources.shaders.volumetricCloudsCombinePS);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs:49]: Missing release

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricCloudsCombine.shader [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            Blend  One SrcAlpha, Zero One
+
+            HLSLPROGRAM
+
+            float4 Frag(Varyings input) : SV_Target
+            {
+                UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(input);
+
+                // Composite the result via hardware blending.
+                return LOAD_TEXTURE2D_X(_VolumetricCloudsUpscaleTextureRW, input.positionCS.xy);;
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricCloudsCombine.shader:59]: extra ;

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricClouds.compute [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #pragma multi_compile _ PHYSICALLY_BASED_SUN
 #pragma multi_compile _ PLANAR_REFLECTION_CAMERA
 #pragma multi_compile _ LOCAL_VOLUMETRIC_CLOUDS
+#pragma multi_compile _ ENABLE_MSAA
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumetricClouds.compute:12]: Why not simply have USE_INTERMEDIATE_BUFFER as the multi compile and add that extra intermediate define?

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs [hunk 4]
```diff
@@ -684,10 +730,11 @@ TextureHandle TraceVolumetricClouds(RenderGraph renderGraph, HDCamera hdCamera,
                 builder.SetRenderFunc(
                     (VolumetricCloudsData data, RenderGraphContext ctx) =>
                     {
-                        TraceVolumetricClouds(ctx.cmd, data.parameters,
+                        TraceVolumetricClouds(ctx.cmd, data.needsTemporaryBuffer, data.cloudCombinePass, data.parameters,
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs:733]: These should be part of the parameters structure

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs [hunk 5]
```diff
@@ -486,12 +491,12 @@ VolumetricCloudsParameters PrepareVolumetricCloudsParameters(HDCamera hdCamera,
             return parameters;
         }
 
-        static void TraceVolumetricClouds(CommandBuffer cmd, VolumetricCloudsParameters parameters,
+        static void TraceVolumetricClouds(CommandBuffer cmd, bool needsTemporaryBuffer, Material combineCloudsPass, VolumetricCloudsParameters parameters,
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/HDRenderPipeline.VolumetricClouds.cs:494]: See other comment

**Review by anisunity — CHANGES_REQUESTED** [review]
The structure is fine, some changes are required

**Review by anisunity — APPROVED** [review]
Looks good to me after the changes!

**Comment by anisunity** [comment]
@johnpars @remi-chapelain we need a new test to cover this!

**Review by sebastienlagarde — APPROVED** [review]

**Comment by anisunity** [comment]
@sebastienlagarde let's not merge until we have a test

**Comment by sebastienlagarde** [comment]
Discussed with John, He will do a separate PR with those test:

A runtime 004 with: volumetric cloud (with planar reflection for cloud), physically based sky, shadow matte, cluster decal (decal on transparent) with decal angle fade + decal with both view bias and depth bias + lens flare when it will be available.
And a runtime 05 with: HDRI sky with distortion layer, SSGI + force forward emissive + area light on hair, fabric, eye, lit + Parallax occlusion mapping with the conservative depth option

**Comment by sebastienlagarde** [comment]
@jans is adding test in a separate PR.

**Pull request merged**